changeset.go

```diff
@@ -109,63 +109,94 @@ func (cs *Changeset) Upsert(ctx context.Context, changesetNamespace, changesetNa
 			}
 			return trace.Wrap(err)
 		}
-		err = cs.upsertResource(ctx, changesetNamespace, changesetName, raw.Raw)
+		err = cs.upsertResource(ctx, changesetNamespace, changesetName, id, raw.Raw)
 		if err != nil {
 			return trace.Wrap(err)
 		}
 	}
 }
 
-func (cs *Changeset) upsertResource(ctx context.Context, changesetNamespace, changesetName string, data []byte) error {
+func (cs *Changeset) upsertResource(ctx context.Context, changesetNamespace, changesetName, id string, data []byte) error {
 	tr, err := cs.createOrRead(changesetNamespace, changesetName, ChangesetSpec{Status: ChangesetStatusInProgress})
 	if err != nil {
 		return trace.Wrap(err)
 	}
 	if tr.Spec.Status != ChangesetStatusInProgress {
 		return trace.CompareFailed("cannot update changeset - expected status %q, got %q", ChangesetStatusInProgress, tr.Spec.Status)
 	}
-	var kind metav1.TypeMeta
-	err = yaml.NewYAMLOrJSONDecoder(bytes.NewReader(data), DefaultBufferSize).Decode(&kind)
+	headers, err := ParseResourceHeader(bytes.NewReader(data))
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	switch kind.Kind {
+
+	comp, err := cs.isOpCompleted(ctx, tr, id)
```

Comment:
Nit: Let's not shorten words unnecessarily: `completed, err := cs.isOperationCompleted(ctx, tr, id)`.